game.py

```diff
@@ -360,6 +393,92 @@ def external_turn(self, action: Action):
                 self._convert_hands(self._obs.hands, self._obs.current_player_id),
             )
 
+    def _discarding_critical_card(self, action: Action, acting_player_id: int) -> int:  
+        """
+        This returns 1 if a player has discarded a critical card
+        """
+        
+        # If action is not a discard, it is irrelevant to the calculation
+        if action.action_type != Action.ActionType.DISCARD:
+            return 0
+
+        # Identify card number and colour
+        (col, num) = self._convert_card(
+            self._obs.hands[acting_player_id][action.cnr]
+        )
+
+        # a 5 card is always critical
+        if num == 5:
+            return 1
+
+        # Check how many instances of the card are in the discard pile
+        trash = self._convert_trash(self._obs.discards)
+        count = 0
+        for card in trash:
+            if card[0] == col and card[1] == num:
+                count += 1
+        
+        # Determine if discard is critical based on card number
+        if num == 1 and count == 2:
+            return 1
+        if (num == 2 or num == 3 or num == 4) and count == 1:
+            return 1
+        
+        return 0
+    
+    def _discarding_known_playable_card(self, action: Action, acting_player_id: int) -> int:
+        """
+        This returns 1 if a player has discarded a known-to-be-playable card
+        """
+        # If action is not a discard, it is irrelevant to the calculation
+        if action.action_type != Action.ActionType.DISCARD:
+            return 0
+        
+        # Identify card number and colour
+        (col, num) = self._convert_card(
+            self._obs.hands[acting_player_id][action.cnr]
+        )
+
+        firework_card = self._convert_board(self._obs.fireworks)[col]
+
+        # Check if the card is playable - only one possible card it could be
+        if num == firework_card[1] + 1:
+            possible_cards = get_possible(self.knowledge[acting_player_id][action.cnr])
+            if len(possible_cards) == 1:
+                return 1
+            
+        return 0
+            
+    def _information_per_play(self, action: Action, acting_player_id: int) -> int:
```

Comment:
The function is annotated to return int, but IPP is a float between 0 and 1 iirc.